File: apps/shared/src/terminal-core-capacity-proto.ts
// Converts terminal-core capacity reports between protobuf bigint fields and the
// shared operational contract. Worker heartbeat and coordinator persistence use
// this one boundary so invalid or lossy reports cannot alter admission visibility.

import { create } from "@bufbuild/protobuf";
import {
  TerminalCoreCapacityReportSchema as TerminalCoreCapacityReportProtoSchema,
  type TerminalCoreCapacityReport as TerminalCoreCapacityReportProto,
} from "./gen/roost/v1/wire_pb.ts";
import {
  TerminalCoreCapacityReportSchema,
  type TerminalCoreCapacityReport,
} from "./terminal-core-capacity.ts";

export function terminalCoreCapacityReportToProto(
  report: TerminalCoreCapacityReport,
): TerminalCoreCapacityReportProto {
  const checked = TerminalCoreCapacityReportSchema.parse(report);
  return create(TerminalCoreCapacityReportProtoSchema, {
    used: checked.used,
    pending: checked.pending,
    capacity: checked.capacity,
    estimatedReservedBytes: BigInt(checked.estimated_reserved_bytes),
    effectiveMemoryCeilingBytes: BigInt(checked.effective_memory_ceiling_bytes),
    bootRssBytes: BigInt(checked.boot_rss_bytes),
    overcommitCount: checked.overcommit_count,
    refusalCount: BigInt(checked.refusal_count),
  });
}

export function terminalCoreCapacityReportFromProto(
  report: TerminalCoreCapacityReportProto,
): TerminalCoreCapacityReport {
  return TerminalCoreCapacityReportSchema.parse({
    used: report.used,
    pending: report.pending,
    capacity: report.capacity,
    estimated_reserved_bytes: boundedNumber(
      report.estimatedReservedBytes,
      "estimated_reserved_bytes",
    ),
    effective_memory_ceiling_bytes: boundedNumber(
      report.effectiveMemoryCeilingBytes,
      "effective_memory_ceiling_bytes",
    ),
    boot_rss_bytes: boundedNumber(report.bootRssBytes, "boot_rss_bytes"),
    overcommit_count: report.overcommitCount,
    refusal_count: boundedNumber(report.refusalCount, "refusal_count"),
  });
}

function boundedNumber(value: bigint, field: string): number {
  if (value > BigInt(Number.MAX_SAFE_INTEGER)) {
    throw new Error(`terminal core capacity ${field} exceeds a safe integer`);
  }
  return Number(value);
}

File: crates/roost-protocol/src/proto_adapters/terminal_core_capacity_proto.rs
//! The terminal core capacity report's protobuf form, which a worker sends to a
//! coordinator deciding viewport geometry. The report itself lives in
//! `wire::worker`; this file only maps it, and runs the report's own
//! cross-field rules in both directions so an inconsistent report cannot alter
//! admission visibility by arriving from a different side of the boundary.
//!
//! The byte counters are `uint64` on both sides, so the safe-integer guard the
//! JavaScript boundary needed before narrowing a `bigint` to a `Number` has
//! nothing left to check here: the whole `u64` range is the field's bound, and a
//! value outside it cannot be expressed at all.

use roost_proto::TerminalCoreCapacityReport as PbTerminalCoreCapacityReport;

use crate::wire::worker::TerminalCoreCapacityReport;
use crate::{ProtocolError, ProtocolResult};

const REPORT: &str = "terminal_core_capacity";

/// Re-check a typed report through the one validator that owns its rules. A
/// Rust caller can assemble a report the cross-field rules would refuse, and the
/// boundary is where the original refused one.
fn checked_report(
    report: &TerminalCoreCapacityReport,
) -> ProtocolResult<TerminalCoreCapacityReport> {
    TerminalCoreCapacityReport::parse(
        serde_json::to_value(report)
            .map_err(|error| ProtocolError::new(REPORT, error.to_string()))?,
    )
}

pub fn terminal_core_capacity_report_to_proto(
    report: &TerminalCoreCapacityReport,
) -> ProtocolResult<PbTerminalCoreCapacityReport> {
    let checked = checked_report(report)?;
    Ok(PbTerminalCoreCapacityReport {
        used: checked.used,
        pending: checked.pending,
        capacity: checked.capacity,
        estimated_reserved_bytes: checked.estimated_reserved_bytes,
        effective_memory_ceiling_bytes: checked.effective_memory_ceiling_bytes,
        boot_rss_bytes: checked.boot_rss_bytes,
        overcommit_count: checked.overcommit_count,
        refusal_count: checked.refusal_count,
        ..Default::default()
    })
}

pub fn terminal_core_capacity_report_from_proto(
    report: &PbTerminalCoreCapacityReport,
) -> ProtocolResult<TerminalCoreCapacityReport> {
    checked_report(&TerminalCoreCapacityReport {
        used: report.used,
        pending: report.pending,
        capacity: report.capacity,
        estimated_reserved_bytes: report.estimated_reserved_bytes,
        effective_memory_ceiling_bytes: report.effective_memory_ceiling_bytes,
        boot_rss_bytes: report.boot_rss_bytes,
        overcommit_count: report.overcommit_count,
        refusal_count: report.refusal_count,
    })
}
